### PR TITLE
feat(oauth): JWT access tokens.

### PR DESCRIPTION
**Pull request by shane-tomlinson**
Super early stage experiment to see how much effort it would be to
convert access tokens to JWTs as defined in https://tools.ietf.org/html/draft-bertocci-oauth-access-token-jwt-00.

Loads of caveats.

JWT isn't signed, still requires DB backing, the code is a mess. You know, normal things.

Hey @rfk, this is ridiculously early stage, but I'm opening it as an exploration because we might need something like this soonish, and I want to see what's possible. It's so early it's not really even ready for concrete feedback, but would appreciate your thoughts anyways.

issue #935

### CODE REVIEW

**Comment by shane-tomlinson** [comment]
Further notes on why I'm exploring this. First, we might need it for some Trailhead stuff. Second, I have this goal of making access tokens more secure, and there is this [Proof-of-Possession Key Semantics for JSON Web Tokens (JWTs) spec](https://tools.ietf.org/html/rfc7800) that might be able to help with that (I'm trying to figure that out), but it requires a JWT.

**Comment by rfk** [comment]
My thoughts can be conveniently summed up by these emoji: 💯 🚀 💥

File: packages/fxa-auth-server/fxa-oauth-server/lib/grant.js [hunk 1]
```diff
@@ -169,3 +172,29 @@ function generateIdToken(grant, access) {
...
+
+  if (grant.amr) {
+    claims.amr = grant.amr;
+  }
+  if (grant.aal) {
+    claims['fxa-aal'] = grant.aal;
+    claims.acr = 'AAL' + grant.aal;
+  }
+
+  // TODO encrypt and sign this thing!
```
Comment (rfk) [inline packages/fxa-auth-server/fxa-oauth-server/lib/grant.js]: Is this as simple as `ID_TOKEN_KEY.sign(claims)`, at least for the first version?  AFAICT from [1] above it's fine to use the same key for id tokens and access tokens (inasmuch as the text says that IdPs *may* elect to use different keys for the two types of token, so it also permits that they use the same key).

**Comment by rfk** [comment]
Another nice outcome of this, will be that we can easily synthesize mock oauth tokens for loadtests.  To do this currently in tokenserver, we currently have to run a little mock OAuth verification service:

  https://github.com/mozilla-services/tokenserver/blob/master/loadtest/mock-oauth-cfn.yml

**Comment by shane-tomlinson** [comment]
> Do you intend to hew to [1] for the contents of the JWT?
> [1] https://datatracker.ietf.org/doc/draft-ietf-oauth-access-token-jwt/

The fields in the JWT come straight [from this section of the spec](https://tools.ietf.org/html/draft-ietf-oauth-access-token-jwt-00#section-2.2).

File: packages/fxa-auth-server/fxa-oauth-server/lib/grant.js (hunk 1 above)
Comment (shane-tomlinson) [reply packages/fxa-auth-server/fxa-oauth-server/lib/grant.js]: It's [recommended that signing be done using asymmetric keys](https://tools.ietf.org/html/draft-ietf-oauth-access-token-jwt-00#section-4), presumably so clients can validate the JWT themselves. If we allowed clients to validate, encrypting the thing doesn't really provide anything extra except to make the token unintelligible to humans. Clients would need to check the signature, then decrypt with the public key. Worthwhile? Not sure. For v1 if we don't allow clients to validate themselves, perhaps.

Comment (rfk) [reply packages/fxa-auth-server/fxa-oauth-server/lib/grant.js]: IIUC `ID_TOKEN_KEY` should already be an asymmetric key.

**Comment by martinthomson** [comment]
Hey @shane-tomlinson, it was suggested that I look at this too, but I don't know your code well enough to really say if this is headed in the right direction.  I was hoping to see some tests here in the hopes of trying to see how this fit together.  Is that the next step?

Let me know if I can help in any way.

**Comment by shane-tomlinson** [comment]
> Hey @shane-tomlinson, it was suggested that I look at this too, but I don't know your code well enough to really say if this is headed in the right direction. I was hoping to see some tests here in the hopes of trying to see how this fit together. Is that the next step?

Thanks @martinthomson, this is a super early stage PR to see whether the approach could work. I can say "it seems to."

The next steps are, like you say, add tests and try to find out whether we missed anything. I have other Trailhead related work too and worry about being a blocker to progress here. Tomorrow I'm going to ask @dannycoates and @vbudhram whether either of them could get this PR across the line. A subsequent PR would be to add the Pairwise Pseudonymous Identifiers.

> Let me know if I can help in any way.

Once this is in a little better shape, your review and suggestions will be highly appreciated.

**Comment by juliemccracken** [comment]
@shane-tomlinson I added this to the SubPlat Support for Secure Proxy epic. Let me know if there's a better place to track this.

**Comment by shane-tomlinson** [comment]
Thanks @juliemccracken, I think it was in the wrong epic before.

**Pull request closed without merging**